File: 24tht_tree.py
def generate_primes(n):
    # Tạo danh sách số nguyên tố sử dụng Sieve of Eratosthenes
    sieve = [True] * (n + 1)
    sieve[0] = sieve[1] = False
    p = 2
    while p * p <= n:
        if sieve[p]:
            for i in range(p * p, n + 1, p):
                sieve[i] = False
        p += 1
    primes = [i for i in range(n + 1) if sieve[i]]
    return primes

def spiral_order(matrix, cannot_plant):
    R, C = len(matrix), len(matrix[0])
    result = []
    seen = [[False] * C for _ in range(R)]
    dr = [0, 1, 0, -1]  # Directions: right, down, left, up
    dc = [1, 0, -1, 0]
    r = c = di = 0  # Starting point and direction index
    for _ in range(R * C):
        if (r, c) not in cannot_plant:
            result.append((r, c))
        seen[r][c] = True
        cr, cc = r + dr[di], c + dc[di]
        if 0 <= cr < R and 0 <= cc < C and not seen[cr][cc]:
            r, c = cr, cc
        else:
            di = (di + 1) % 4  # Change direction
            r, c = r + dr[di], c + dc[di]
    return result


# Input processing
N, M, K, Q = map(int, input().split())
cannot_plant = set()
for _ in range(K):
    x, y = map(int, input().split())
    cannot_plant.add((x - 1, y - 1))  # Convert to 0-indexed

primes = generate_primes(int(4e6))  # Generate enough primes
grid = [[0] * M for _ in range(N)]
spiral = spiral_order(grid, cannot_plant)
prime_index = 0
for r, c in spiral:
    if (r, c) not in cannot_plant:
        grid[r][c] = primes[prime_index]
        prime_index += 1
pref = [[0] * (M + 5) for _ in range(N + 5)]  # Prefix sum of grid values
for i in range(1, N + 1):
    for j in range(1, M + 1):
        pref[i][j] = pref[i - 1][j] + pref[i][j - 1] - pref[i - 1][j - 1] + grid[i - 1][j - 1]

# Process each query
results = []
for _ in range(Q):
    x1, y1, x2, y2 = map(int, input().split())
    results.append(pref[x2][y2] - pref[x1 - 1][y2] - pref[x2][y1 -1] + pref[x1 - 1][y1 - 1])

# Output results
for result in results:
    print(result)

File: 24tht_tree.cpp
#include <iostream>
#include <vector>
#include <set>
using namespace std;

vector<int> generate_primes(int n) {
    vector<bool> sieve(n + 1, true);
    sieve[0] = sieve[1] = false;
    for (int p = 2; p * p <= n; ++p) {
        if (sieve[p]) {
            for (int i = p * p; i <= n; i += p) {
                sieve[i] = false;
            }
        }
    }
    vector<int> primes;
    for (int i = 0; i <= n; ++i) {
        if (sieve[i]) {
            primes.push_back(i);
        }
    }
    return primes;
}

vector<pair<int, int>> spiral_order(int R, int C, const set<pair<int, int>>& cannot_plant) {
    vector<pair<int, int>> result;
    vector<vector<bool>> seen(R, vector<bool>(C, false));
    int dr[4] = {0, 1, 0, -1}; // Directions: right, down, left, up
    int dc[4] = {1, 0, -1, 0};
    int r = 0, c = 0, di = 0; // Starting point and direction index
    for (int i = 0; i < R * C; ++i) {
        if (cannot_plant.find({r, c}) == cannot_plant.end()) {
            result.push_back({r, c});
        }
        seen[r][c] = true;
        int cr = r + dr[di], cc = c + dc[di];
        if (0 <= cr && cr < R && 0 <= cc && cc < C && !seen[cr][cc]) {
            r = cr;
            c = cc;
        } else {
            di = (di + 1) % 4; // Change direction
            r += dr[di];
            c += dc[di];
        }
    }
    return result;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);
    int N, M, K, Q;
    cin >> N >> M >> K >> Q;
    set<pair<int, int>> cannot_plant;
    for (int i = 0; i < K; ++i) {
        int x, y;
        cin >> x >> y;
        cannot_plant.insert({x - 1, y - 1}); // Convert to 0-indexed
    }

    vector<int> primes = generate_primes(4000000); // Generate enough primes
    vector<vector<int>> grid(N, vector<int>(M, 0));
    vector<pair<int, int>> spiral = spiral_order(N, M, cannot_plant);
    int prime_index = 0;
    for (const auto& p : spiral) {
        int r = p.first, c = p.second;
        if (cannot_plant.find({r, c}) == cannot_plant.end()) {
            grid[r][c] = primes[prime_index++];
        }
    }

    vector<vector<long long>> pref(N + 5, vector<long long>(M + 5, 0)); // Prefix sum of grid values
    for (int i = 1; i <= N; ++i) {
        for (int j = 1; j <= M; ++j) {
            pref[i][j] = pref[i - 1][j] + pref[i][j - 1] - pref[i - 1][j - 1] + grid[i - 1][j - 1];
        }
    }

    // Process each query
    vector<long long> results;
    for (int i = 0; i < Q; ++i) {
        int x1, y1, x2, y2;
        cin >> x1 >> y1 >> x2 >> y2;
        results.push_back(pref[x2][y2] - pref[x1 - 1][y2] - pref[x2][y1 - 1] + pref[x1 - 1][y1 - 1]);
    }

    // Output results
    for (const auto& result : results) {
        cout << result << endl;
    }

    return 0;
}